src/shogun/distributions/KernelDensity.cpp

```diff
@@ -91,12 +91,14 @@ bool KernelDensity::train(std::shared_ptr<Features> data)
 	return true;
 }
 
-SGVector<float64_t> KernelDensity::get_log_density(const std::shared_ptr<DenseFeatures<float64_t>>& test, int32_t leaf_size)
+SGVector<float64_t> KernelDensity::get_log_density(const std::shared_ptr<Features>& test, int32_t leaf_size)
 {
 	require(test,"data not supplied");
-
+	require(test->get_feature_type == 10 && test->get_feature_class == 11,"Expecting DenseFeatures<float64_t>");
+	std::shared_ptr<DenseFeatures<float64_t>> dense_feat = std::static_pointer_cast<DenseFeatures<float64_t>>(test);
```

Comment:
imo you should do a dynamic cast and check for nullptr, you shouldn't rely on `get_feature_type` that might get deprecated soon.

Reply:
Aight I'll update that

Reply:
I've used ```dynamic_pointer_cast``` as it's ```std::shared_ptr``` thoughts?

Reply:
yes that's what you should use